cmd/crank/beta/trace/trace.go

```diff
@@ -0,0 +1,143 @@
+	errCannotGetKindAndName   = "cannot get kind and name"
+	errCannotGetMapping       = "cannot get mapping for resource"
+	errCannotInitPrinter      = "cannot init new printer"
+	errFmtInvalidResourceName = "invalid combined kind and name format, should be in the form of 'resource.group.example.org/name', got: %q"
+)
+
+// Cmd builds the trace tree for a Crossplane resource.
+type Cmd struct {
+	Resource string `arg:"" required:"" help:"'TYPE[.VERSION][.GROUP][/NAME]' identifying the Crossplane resource."`
+	Name     string `arg:"" optional:"" help:"Name of the Crossplane resource. Ignored if already passed as part of the RESOURCE argument."`
+
+	// TODO(phisco): add support for all the usual kubectl flags; configFlags := genericclioptions.NewConfigFlags(true).AddFlags(...)
+	// TODO(phisco): move to namespace defaulting to "" and use the current context's namespace
+	Namespace string `short:"n" name:"namespace" help:"Namespace of resource." default:"default"`
+	Output    string `short:"o" name:"output" help:"Output format. One of: default, json, dot." enum:"default,json,dot" default:"default"`
+}
+
+// Run runs the trace command.
+func (c *Cmd) Run(k *kong.Context, logger logging.Logger) error { //nolint:gocyclo // TODO(phisco): refactor
+	logger = logger.WithValues("Resource", c.Resource)
+
+	// Init new printer
+	p, err := printer.New(c.Output)
+	if err != nil {
+		return errors.Wrap(err, errCannotInitPrinter)
+	}
+	logger.Debug("Built printer", "output", c.Output)
+
+	kubeconfig, err := ctrl.GetConfig()
+	if err != nil {
+		logger.Debug(errKubeConfig, "error", err)
+		return errors.Wrap(err, errKubeConfig)
+	}
+	logger.Debug("Found kubeconfig")
+
+	client, err := client.New(kubeconfig, client.Options{
+		Scheme: scheme.Scheme,
+	})
+	if err != nil {
+		return errors.Wrap(err, errCouldntInitKubeClient)
+	}
+
+	var opts []resource.ClientOption
+	if p.NeedLatestEvent() {
+		logger.Debug("Adding latest event option")
+		opts = append(opts, resource.WithLatestEvent())
+	}
+
+	// Get client for k8s package
+	resClient, err := resource.NewClient(client, opts...)
+	if err != nil {
+		return errors.Wrap(err, errCouldntInitKubeClient)
+	}
+	logger.Debug("Built client")
+
+	kind, name, err := c.getKindAndName()
+	if err != nil {
+		return errors.Wrap(err, errCannotGetKindAndName)
+	}
+
+	mapping, err := resClient.MappingFor(kind)
+	if err != nil {
+		return errors.Wrap(err, errCannotGetMapping)
+	}
+
+	// Get Resource object. Contains k8s resource and all its children, also as Resource.
+	rootRef := &v1.ObjectReference{
+		Kind:       mapping.GroupVersionKind.Kind,
+		APIVersion: mapping.GroupVersionKind.GroupVersion().String(),
+		Name:       name,
+	}
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace && c.Namespace != "" {
+		rootRef.Namespace = c.Namespace
+	}
+	logger.Debug("Getting resource tree", "rootRef", rootRef.String())
+	root, err := resClient.GetResourceTree(context.Background(), rootRef)
```

Comment:
I'm think this `GetResourceTree()` call is failing for my scenario in a way that I would not expect it to:

```
❯ ./crank beta trace acmedatabases.v1.acme.com/acme-db-prod
crossplane: error: cannot get requested resource: couldn't get child resource: couldn't get resource: secrets "acme-db-prod-conn" not found
```

Basically my scenario here is creating a claim but forgetting to provide credentials to the cloud provider. So I'm expecting this claim to not become ready and I want to find out why with this `trace` command.

The connection secret for my claim hasn't been created yet, but it looks like this function doesn't handle that gracefully in its effort to discover the entire graph. It doesn't find the connection secret (as expected, it doesn't exist) and then returns an error for that instead of continuing to trace through the rest of the graph.

It would probably be a better experience to handle the missing connection secret and still print out for me the graph of resources so I can visualize them and see where the root failure is.

Some more debugging information:

General claim output: 

```
❯ k get acmedatabases.v1.acme.com/acme-db-prod
NAME           SYNCED   READY   CONNECTION-SECRET   AGE
acme-db-prod   True     False   acme-db-prod-conn   8m51s
```

`--verbose` output:

```
❯ ./crank beta trace acmedatabases.v1.acme.com/acme-db-prod --verbose
2023-10-21T07:51:38-07:00	DEBUG	Built printer	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod", "output": "default"}
2023-10-21T07:51:38-07:00	DEBUG	Found kubeconfig	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod"}
2023-10-21T07:51:38-07:00	DEBUG	Adding latest event option	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod"}
2023-10-21T07:51:38-07:00	DEBUG	Built client	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod"}
2023-10-21T07:51:38-07:00	DEBUG	Getting resource tree	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod", "rootRef": "&ObjectReference{Kind:AcmeDatabase,Namespace:default,Name:acme-db-prod,UID:,APIVersion:acme.com/v1,ResourceVersion:,FieldPath:,}"}
2023-10-21T07:51:38-07:00	DEBUG	cannot get requested resource	{"Resource": "acmedatabases.v1.acme.com/acme-db-prod", "error": "couldn't get child resource: couldn't get resource: secrets \"acme-db-prod-conn\" not found"}
crossplane: error: cannot get requested resource: couldn't get child resource: couldn't get resource: secrets "acme-db-prod-conn" not found
```

The relationship of claim -> XR -> MR that I was hoping to see output, even though the connection secret doesn't exist yet in the claim:

```
❯ k get acmedatabase acme-db-prod -o json | jq '.spec.resourceRef'
{
  "apiVersion": "acme.com/v1",
  "kind": "XAcmeDatabase",
  "name": "acme-db-prod-7sc7m"
}

❯ k get xacmedatabase acme-db-prod-7sc7m -o json | jq '.spec.resourceRefs'
[
  {
    "apiVersion": "sql.gcp.upbound.io/v1beta1",
    "kind": "DatabaseInstance",
    "name": "acme-db-prod-7sc7m-2ggn9"
  }
]

❯ k get databaseinstance acme-db-prod-7sc7m-2ggn9 -o json | jq '.status'
{
  "atProvider": {},
  "conditions": [
    {
      "lastTransitionTime": "2023-10-21T14:38:18Z",
      "message": "connect failed: cannot get terraform setup: cannot extract JSON key credentials: cannot get credentials secret: Secret \"gcp-secret\" not found",
      "reason": "ReconcileError",
      "status": "False",
      "type": "Synced"
    }
  ]
}
```

Reply:
Good point! Properly handling expected resources not existing is definitely something we want to show and highlight as a possible source of issue. Any suggestion about how to graphically show that in the tree view? An additional boolean FOUND column? Or a generic ERROR column where we could show errors for each resource? I think the latter could be more flexible.

Reply:
Hmm actually I don't (yet) think it's important to display this particular scenario with any specific knowledge/logic.  The missing connection secret for the claim is a normal part of its lifecycle - it hasn't gotten far enough for the connection secret to be created. So I wouldn't elevate that symptom to a top level experience personally.

What I would do though is to not let that temporary condition stop the processing of the rest of the graph.  In this scenario, where we really want to get to is the event or condition at the MR - that's where the real root cause is here.

So just by displaying the entire tree/graph of claim --> XR --> MR and showing the last event or status or something on the MR would be enough for this scenario.  No need for extra columns or experiences here is my bet! 🙇

Reply:
took the latest for another test spin, I was able to get the full output of the tree graph now! the root cause is printed within the output, so this is definitely helpful 💪 

```
❯ ./crank beta trace acmedatabases acme-db-prod
NAMESPACE           APIVERSION                   NAME                                                                 READY     SYNCED   STATUS
default             acme.com/v1                  AcmeDatabase/acme-db-prod                                            False     True     Waiting: "Composite resource claim is waiting for composite resource to become Ready"
                    acme.com/v1                  ├── XAcmeDatabase/acme-db-prod-7sc7m                                 False     True     Creating: ""
                    sql.gcp.upbound.io/v1beta1   │   ├── DatabaseInstance/acme-db-prod-7sc7m-2ggn9                    Unknown   False    ReconcileError: "connect failed: cannot get terraform setup: cannot extract JSON key credentials: cannot get credentials secret: Secret \"gcp-secret\" not found"
crossplane-system   v1                           │   │   └── Secret/5d2b9592-d3c7-4dc5-b27c-67a3c10c1b45-postgresql                      Error: secrets "5d2b9592-d3c7-4dc5-b27c-67a3c10c1b45-postgresql" not found
crossplane-system   v1                           │   └── Secret/5d2b9592-d3c7-4dc5-b27c-67a3c10c1b45                                     : ""
default             v1                           └── Secret/acme-db-prod-conn                                                            Error: secrets "acme-db-prod-conn" not found
```

Further thoughts on this output:

* **wrapping by default**: with the default font/zoom size of my terminal on my laptop screen, the output is wide enough to wrap some lines, which causes breaks in the tree lines and makes the valuable visualization of the tree much more difficult to parse. Thoughts on limiting the width of each line somehow?
* **namespace and apiversion columns** i'm wondering how valuable these columns are in the default view, esp. when the namespace column is often blank because of cluster scoped XRs and MRs. They are taking up almost 30% of the horizontal real estate - what do you think about not including them in the default view? I keep coming back to thinking the `NAME` column should be first so it sort of anchors the graph on the left hand side.
* **connection secrets** I was surprised to see connection secrets included in the graph and they don't seem to be helpful for finding root cause here either. I'm not sure how much valuable details they would be able to show in this graph since they don't have a READY/SYNCED value and they probably don't have a useful STATUS value either. I was imagining including only resources that are XRM compliant such as claims, XRs, and MRs.  Perhaps we can have a flag to `--show-secrets` for folks to optionally include them in the output, but have them excluded by default?

Reply:
Totally agree with all your points, I've:
- cropped the condition message to 64 characters, the full output can be shown by using `-o wide` (we could think of additional columns later, as kubectl does)
- added a `--show-connection-secrets` flag, false by default
- dropped namespace and apiversion columns, namespace if present is added at the end of the name between brackets. The apiVersion could be useful in case of multiple colliding types, could potentially be a column we add to the wide output.

Reply:
otherwise we could just drop the message part from the status similarly to what kubectl get does and leave getting that to further investigation 🤔

Reply:
Did you push those latest updates to your branch yet @phisco? i'd love to see that code and give those changes a test spin too! 🙇‍♂️ 

I'd be hesitant to remove the message part of the status because in my main testing scenario, and other unhealthy scenarios I can imagine, the message is going to have the most useful information about the root cause. Let's keep it until we arrive at a conclusion that it's causing more harm than good - my expectation is that it will cause lots of good 😊 

How'd you arrive at 64 chars by the way? Do you think we'd have room for more than that since the namespace/apiversion colums are now gone?

Reply:
Just pushed, I forgot to do that earlier, let me know!

64 is just a random number that worked well in my scenario, we could probably compute that based on the terminal width, but I'm not sure it's worth it.

With wrapped errors the most meaningful part is at the end of the message so I just left the full condition status, e.g. Waiting, and then cut the message to just the last 64 characters, printing it in full with wide output.

E.g.

```
NAME                                                       READY     SYNCED   STATUS
ANopResource/apiextensions-composition-minimal (default)   False     True     Waiting: ...resource claim is waiting for composite resource to become Ready
└─ XANopResource/apiextensions-composition-minimal-2w2lr   Unknown   False    ReconcileError: cannot select Composition: no compatible Compositions found
```

Reply:
Awesome dude, just took it for a real quick test spin and it looks really good dude, I think we're landing on what we'd like to ship for v1.14!!

I'll take a deeper code review pass this afternoon too so hopefully we'll be able to merge first thing Monday 😅

Reply:
Awesome, thanks! I could definitely add some more test coverage and the default printer's code could benefit a cleanup/comment pass for sure. I'll do that tomorrow morning, while integrating your feedback 🙏